allennlp/modules/seq2vec_encoders/wrapped_pytorch_rnn.py

```diff
@@ -0,0 +1,53 @@
+from typing import Type
+
+import torch
+
+from allennlp.common import Params
+from allennlp.common.checks import ConfigurationError
+from allennlp.modules import Seq2VecEncoder
+
+
+class WrappedPytorchRnn(Seq2VecEncoder):
+    """
+    Pytorch's RNNs have two outputs: the hidden state for every time step, and the hidden state at
+    the last time step for every layer.  We just want the second one as a single output.  This
+    wrapper pulls out that output, and allows us to instantiate the pytorch RNN from parameters.
+
+    This is actually slightly more complicated than that, because the LSTM's second output is a
+    tuple of (hidden state, cell state), and we just return the hidden state.
+
+    Also, there are lots of ways you could imagine going from an RNN hidden state at every
+    timestep to a single vector - you could take the last vector at all layers in the stack, do
+    some kind of pooling, take the last vector of the top layer in a stack, or many other  options.
+    We just take the final hidden state vector.  TODO(mattg): allow for other ways of wrapping
+    RNNs.
+    """
+
+    def __init__(self, module: torch.nn.modules.RNNBase) -> None:
+        super(WrappedPytorchRnn, self).__init__()
+        self._module = module
+
+    def get_output_dim(self) -> int:
+        return self._module.hidden_size * self._module.num_directions
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:  # pylint: disable=arguments-differ
+        return self._module(inputs)[0][-1]
+
+    class _Wrapper:
```

Comment:
There are a few issues here:

1. I was trying to not have the `_Wrapper` show up in the module namespace.  That's why this is nested.  But, it's probably more confusing than it's worth to do this, so I can just move the `_Wrapper` class into the module init.  It should be a lot more clear what's going on.

2. The registry dictionaries are supposed to be `Dict[str, Type[Something]]`, where `Something` is typically either initialized directly, with `Something()`, or `from_params`, with `Something.from_params`.  So the thing that goes into the dict has to have a `__call__` method that mimics `Something.__init__()`, and a `from_params` method that mimics `Something.from_params`.  We can't just return a `Something` object because `Something` is not `Type[Something]`, and we can't return `WrappedPytorchRnn` because this class needs to be instantiated with an instantiated pytorch RNN, which breaks the constructor and `from_params` API assumptions.  I'll add some of this info to the docstring for `_Wrapper` when I move it to the module `__init__.py`.

3. I could add a `from_params` method directly to the `WrappedPytorchRnn`, but I think it's better to omit it and use the `_Wrapper`.  Here's the practical significance of that decision:

With `_Wrapper`:

```
"encoder": {
  "type": "lstm",
  "input_dim": 100,
  ...  // other JSON-encodable pytorch LSTM parameters here
}
```

Without `_Wrapper`:

```
"encoder": {
  "type": "wrapped",
  "wrapped_type": "lstm",
  "input_dim": 100,
  ...  // other JSON-encodable pytorch LSTM parameters here
}
```

And then `WrappedPytorchRnn` has to know about mapping strings to pytorch RNN classes, instead of just having a single mapping in the module init.

4. Would you prefer `WrappedPytorchSeq2{Seq,Vec}Encoder`, or similar, instead of having these wrapper classes have identical names, but different packages?  If you ever want to import both of them into the same class, keeping them with the same name could be confusing, but the idea is that you wouldn't ever be importing these directly, anyway - if you're to the point of importing encoders and not using our params stuff, you're probably better off just importing `LSTM` directly from pytorch, anyway.  I'll add a note in the docstrings of these classes to that effect.